packages/headless/doc-parser/src/extractor.ts

```diff
@@ -0,0 +1,59 @@
+import {AnyEntity, isObjectEntity, ObjEntity} from './entity';
+import {buildEntity, buildObjEntity} from './entity-builder';
+
+interface Extraction {
+  types: ObjEntity[];
+}
+
+export function extractTypes(headingEntities: AnyEntity[]): Extraction {
+  const extraction: Extraction = {
+    types: [],
+  };
+
+  processObjectEntities(headingEntities, extraction, 0);
+
+  return extraction;
+}
+
+function processObjectEntities(
+  entities: AnyEntity[],
+  extraction: Extraction,
+  level: number
+) {
+  entities
+    .filter(isObjectEntity)
+    .forEach((entity) => extract(entity, extraction, level + 1));
+}
+
+function extract(entity: ObjEntity, extraction: Extraction, level: number) {
+  const maxDepth = 2;
+  const {members} = entity;
+
+  if (!members.length) {
+    return;
+  }
+
+  if (level === maxDepth) {
+    const typeHeading = buildTypeEntity(entity, members);
+    extraction.types.push(typeHeading);
+
+    entity.isTypeExtracted = true;
+    entity.members = [];
+
+    processObjectEntities([typeHeading], extraction, 0);
+    return;
+  }
+
+  processObjectEntities(members, extraction, level);
+}
+
+function buildTypeEntity(originalEntity: ObjEntity, members: AnyEntity[]) {
+  const entity = buildEntity({
+    name: originalEntity.typeName,
+    type: '',
+    isOptional: false,
+    comment: undefined,
```

Comment:
Since a decent amount of properties seem to be unnecessary here, should a new type such as `ExtractedObjEntity` or something of that sort be created? I feel like `ObjEntity` is being used for too many purposes.

Reply:
Really good comment. I explored this a little, but felt introducing a new entity type was making the function signatures in this file more complex.

For example, the `extract` function would now need to have signature
```
function extract(entity: ObjEntity | ExtractedObjEntity) {}
```
and `processObjectEntities` would become `processObjectAndExtractedObjectEntites` or `processEntitiesWithMembers`.

I agree filling out unnecessary properties is not the cleanest though. But I also feel doing things correctly will add complexity compared to keeping one "shape", hmm.

What do you think? Maybe I am missing something?